### PR TITLE
NIFI-11430 - Fixed empty file being produced when grouping by query and fixed _source and _meta extraction by query

### PR DESCRIPTION
**Pull request by r-vandenbos**
# Summary

[NIFI-11430](https://issues.apache.org/jira/browse/NIFI-11430)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 11
  - [ ] JDK 17

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Comment by exceptionfactory** [comment]
Thanks for the contribution @r-vandenbos!

One procedural note, it looks like the username on this pull request @r-vandenbos does not match the username on the Git commit, which is @vandenbos. In general, these usernames should match for proper attribution of both the PR and the commits. It should be possible to update the author on the commit to match.

**Comment by r-vandenbos** [comment]
@exceptionfactory thanks for pointing this out, I got into a spot a trouble with local vs global settings.

This has been fixed now

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/PaginatedJsonQueryElasticsearchTest.groovy [hunk 1]
```diff
@@ -61,6 +61,21 @@ class PaginatedJsonQueryElasticsearchTest extends AbstractPaginatedJsonQueryElas
         reset(runner)
 
 
+        // NIFI-11430 test that 1 flow file is output when hits output per response with output empty flow file set to true
+        runner.setProperty(AbstractJsonQueryElasticsearch.OUTPUT_NO_HITS, "true")
```
Comment (ChrisSamo632) [inline nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/PaginatedJsonQueryElasticsearchTest.groovy:65]: Once a property has been set for the processor on the `runner`, it doesn't get reset/changed unless explicitly modified by the test code - the existing tests were presumably using the default of not outputting `no_hits`, the new tests are setting this to `true` and leaving it that way with little/no other assertion changes (i.e. we basically run the exact same test and aseertions twice)

Be sure to set `runner.setProperty(AbstractJsonQueryElasticsearch.OUTPUT_NO_HITS, "false")` for each of the existing sets of tests here before the `runOnce(runner)` calls

Even better, consider refactoring to loop through the different values (`OUTPUT_NO_HITS`, but possibly `SEARCH_RESULTS_SPLIT` too), similar to how you've done it in `AbstractPaginatedJsonQueryElasticsearchTest.groovy`

**Review by ChrisSamo632 — CHANGES_REQUESTED** [review]
Thanks for the contribution @r-vandenbos, a few minor comments from me

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearch.java [hunk 2]
```diff
@@ -264,14 +265,16 @@ private void combineHits(final List<Map<String, Object>> hits, final PaginatedJs
...
                               final ProcessSession session, final FlowFile parent, final Map<String, String> attributes,
                               final List<FlowFile> hitsFlowFiles, final String transitUri, final StopWatch stopWatch) throws IOException {
         paginatedJsonQueryParameters.incrementPageCount();
         attributes.put("page.number", Integer.toString(paginatedJsonQueryParameters.getPageCount()));
 
         if (hitStrategy == ResultOutputStrategy.PER_QUERY) {
-            combineHits(hits, paginatedJsonQueryParameters, session, parent, attributes, hitsFlowFiles);
+
+            hits = formatHits(hits);
+            combineHits(hits, paginatedJsonQueryParameters, session, parent, attributes, hitsFlowFiles, newQuery);
```
[the remark this answers is not in the document]
Comment (r-vandenbos) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearch.java]: 👍🏻

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearch.java [hunk 3]
```diff
@@ -264,14 +265,16 @@ private void combineHits(final List<Map<String, Object>> hits, final PaginatedJs
      * SearchResponse is processed, i.e. this approach allows recursion for paginated queries, but is unnecessary for single-response queries.
      */
     @Override
-    List<FlowFile> handleHits(final List<Map<String, Object>> hits, final boolean newQuery, final PaginatedJsonQueryParameters paginatedJsonQueryParameters,
+    List<FlowFile> handleHits(List<Map<String, Object>> hits, final boolean newQuery, final PaginatedJsonQueryParameters paginatedJsonQueryParameters,
```
[the remark this answers is not in the document]
Comment (r-vandenbos) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearch.java]: 👍🏻

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearchTest.groovy [hunk 4]
```diff
@@ -32,15 +32,15 @@ import static groovy.json.JsonOutput.toJson
 import static org.hamcrest.CoreMatchers.equalTo
 import static org.hamcrest.CoreMatchers.is
 import static org.hamcrest.MatcherAssert.assertThat
-import static org.junit.jupiter.api.Assertions.assertThrows
+import static org.junit.jupiter.api.Assertions.*
```
[the remark this answers is not in the document]
Comment (r-vandenbos) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearchTest.groovy]: 👍🏻

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearchTest.groovy [hunk 5]
```diff
@@ -117,13 +117,184 @@ abstract class AbstractPaginatedJsonQueryElasticsearchTest extends AbstractJsonQ
...
+                    break
+            }
+
+            // test _source only format
+            runner.setProperty(AbstractJsonQueryElasticsearch.SEARCH_RESULTS_SPLIT, resultOutputStrategy.getValue())
+            runner.setProperty(AbstractJsonQueryElasticsearch.SEARCH_RESULTS_FORMAT, SearchResultsFormat.SOURCE_ONLY.getValue())
+
+
+            // Test against each pagination type
+            for (PaginationType paginationType : PaginationType.values()) {
```
[the remark this answers is not in the document]
Comment (r-vandenbos) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearchTest.groovy]: 👍🏻

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearchTest.groovy [hunk 6]
```diff
@@ -117,13 +117,184 @@ abstract class AbstractPaginatedJsonQueryElasticsearchTest extends AbstractJsonQ
         assertSendEvent(runner, input)
     }
 
+    @Test
+    void testSourceExtraction() throws Exception {
+        for (ResultOutputStrategy resultOutputStrategy : ResultOutputStrategy.values()) {
```
[the remark this answers is not in the document]
Comment (r-vandenbos) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearchTest.groovy]: 👍🏻

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearchTest.groovy [hunk 7]
```diff
@@ -117,13 +117,184 @@ abstract class AbstractPaginatedJsonQueryElasticsearchTest extends AbstractJsonQ
...
+                    break
+                case ResultOutputStrategy.PER_HIT:
+                    flowFileCount = 10
+                    hitsCount = "1"
+                    break
+                case ResultOutputStrategy.PER_RESPONSE:
+                    flowFileCount = 1
+                    hitsCount = "10"
+                    break
+            }
```
[the remark this answers is not in the document]
Comment (r-vandenbos) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearchTest.groovy]: 👍🏻

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearchTest.groovy [hunk 8]
```diff
@@ -265,46 +436,35 @@ abstract class AbstractPaginatedJsonQueryElasticsearchTest extends AbstractJsonQ
...
-        // test that an empty flow file is produced for a per query setup
-        runner.setProperty(AbstractPaginatedJsonQueryElasticsearch.SEARCH_RESULTS_SPLIT, ResultOutputStrategy.PER_QUERY.getValue())
-        runOnce(runner)
-        testCounts(runner, isInput() ? 1 : 0, 1, 0, 0)
 
-        runner.getFlowFilesForRelationship(AbstractJsonQueryElasticsearch.REL_HITS).get(0).assertAttributeEquals("hit.count", "0")
-        runner.getFlowFilesForRelationship(AbstractJsonQueryElasticsearch.REL_HITS).get(0).assertAttributeEquals("page.number", "1")
-        runner.getFlowFilesForRelationship(AbstractJsonQueryElasticsearch.REL_HITS).get(0).getSize() == 0
-        reset(runner)
+        for (PaginationType paginationType : PaginationType.values()) {
```
[the remark this answers is not in the document]
Comment (r-vandenbos) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearchTest.groovy]: 👍🏻

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearchTest.groovy [hunk 9]
```diff
@@ -265,46 +436,35 @@ abstract class AbstractPaginatedJsonQueryElasticsearchTest extends AbstractJsonQ
...
+            runner.setProperty(AbstractPaginatedJsonQueryElasticsearch.PAGINATION_TYPE, paginationType.getValue())
 
-        // test that an empty flow file is produced for a per hit setup
-        runner.setProperty(AbstractPaginatedJsonQueryElasticsearch.SEARCH_RESULTS_SPLIT, ResultOutputStrategy.PER_HIT.getValue())
-        runOnce(runner)
-        testCounts(runner, isInput() ? 1 : 0, 1, 0, 0)
+            for (ResultOutputStrategy resultOutputStrategy : ResultOutputStrategy.values()) {
```
[the remark this answers is not in the document]
Comment (r-vandenbos) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearchTest.groovy]: 👍🏻

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearchTest.groovy [hunk 10]
```diff
@@ -117,13 +117,184 @@ abstract class AbstractPaginatedJsonQueryElasticsearchTest extends AbstractJsonQ
...
+                    })
+                }
+
+                reset(runner)
+            }
+        }
+    }
+
+    @Test
+    void testMetaExtraction() throws Exception {
```
[the remark this answers is not in the document]
Comment (r-vandenbos) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/AbstractPaginatedJsonQueryElasticsearchTest.groovy]: 👍🏻

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/PaginatedJsonQueryElasticsearchTest.groovy (hunk 1 above)
Comment (r-vandenbos) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/groovy/org/apache/nifi/processors/elasticsearch/PaginatedJsonQueryElasticsearchTest.groovy]: 👍🏻

**Review by ChrisSamo632 — APPROVED** [review]
Thanks for the update @r-vandenbos, I've kicked off the GitHub Acitons and provided they pass, this change LGTM 👍

**Pull request closed without merging**